### DISCUSSION TITLE
How can I send any msgs in studio publish panel, it tell me that "Error: Tried to publish on topic '/chat_to_server' that has not been advertiesd before"

### DISCUSSION THREAD

**Discussion by chaosLeee** [foxglove/community#809] [Q&A] [answered]
what i should do in server to accpet this topic, is there any docs for help?

**Comment by chaosLeee** [comment]
i want to know which callback should i handle in foxglove websocket server to support this function? clientAdvertiseHandler, clientUnadvertiseHandler ,clientMessageHandler?

**Reply by achim-k** [reply, marked as answer]
The error you are seeing [comes from Studio](https://github.com/foxglove/studio/blob/62129b44b9e2eb95adc8f8656ebcc5aa57af7c0a/packages/studio-base/src/players/FoxgloveWebSocketPlayer/index.ts#L959-L962). It means that your client channel could not be created either due to
- The server [not supporting one of Studio's supported encodings](https://github.com/foxglove/studio/blob/62129b44b9e2eb95adc8f8656ebcc5aa57af7c0a/packages/studio-base/src/players/FoxgloveWebSocketPlayer/index.ts#L1111-L1126)
- The[ lack of datatypes when publishing with one of the ROS encodings](https://github.com/foxglove/studio/blob/62129b44b9e2eb95adc8f8656ebcc5aa57af7c0a/packages/studio-base/src/players/FoxgloveWebSocketPlayer/index.ts#L1129-L1150)

I suggest that you take a look at the example [Python JSON-Server](https://github.com/foxglove/ws-protocol/blob/main/python/src/foxglove_websocket/examples/json_server.py) and take that as an example for writing your own server.
